// @asymmetrik/ngx-leaflet-draw/__ivy_ngcc__/dist/leaflet-draw/leaflet-draw.module.js
import { NgModule } from '@angular/core';
import { LeafletModule } from '@asymmetrik/ngx-leaflet';
import { LeafletDrawDirective } from './core/leaflet-draw.directive';
import * as ɵngcc0 from '@angular/core';
var LeafletDrawModule = /** @class */ (function () {
    function LeafletDrawModule() {
    }
LeafletDrawModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LeafletDrawModule });
LeafletDrawModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LeafletDrawModule_Factory(t) { return new (t || LeafletDrawModule)(); }, imports: [[
            LeafletModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LeafletDrawModule, { declarations: function () { return [LeafletDrawDirective]; }, imports: function () { return [LeafletModule]; }, exports: function () { return [LeafletDrawDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LeafletDrawModule, [{
        type: NgModule,
        args: [{
                imports: [
                    LeafletModule
                ],
                exports: [
                    LeafletDrawDirective
                ],
                declarations: [
                    LeafletDrawDirective
                ]
            }]
    }], function () { return []; }, null); })();
    return LeafletDrawModule;
}());
export { LeafletDrawModule };
